jmx_prometheus_javaagent/src/main/java/io/prometheus/jmx/JavaAgent.java

```diff
@@ -20,37 +22,71 @@ public static void premain(String agentArgument, Instrumentation instrumentation
-     }
-
-     int port;
-     String file;
-     InetSocketAddress socket;
-
-     if (args.length == 3) {
-       port = Integer.parseInt(args[1]);
-       socket = new InetSocketAddress(args[0], port);
-       file = args[2];
-     } else {
-       port = Integer.parseInt(args[0]);
-       socket = new InetSocketAddress(host, port);
-       file = args[1];
-     }
+     Config config = parseConfig(agentArgument, host);
 
      new BuildInfoCollector().register();
-     new JmxCollector(new File(file)).register();
+     new JmxCollector(new File(config.file)).register();
      DefaultExports.initialize();
-     server = new HTTPServer(socket, CollectorRegistry.defaultRegistry, true);
+     server = new HTTPServer(config.socket, CollectorRegistry.defaultRegistry, true);
    }
+
+    /**
+     * Parse the Java Agent configuration. The arguments are typically specified to the JVM as a javaagent as
+     * {@code -javaagent:/path/to/agent.jar=<CONFIG>}. This method parses the {@code <CONFIG>} portion.
+     * @param args provided agent args
+     * @param ifc default bind interface
+     * @return configuration to use for our application
+     */
+    public static Config parseConfig(String args, String ifc) {
+        // If we have IPv6 address in square brackets, extract it first and then
+        // remove it from arguments to prevent confusion from too many colons.
+        Integer indexOfClosingSquareBracket = args.indexOf("]:");
+        if (indexOfClosingSquareBracket >= 0) {
+            ifc = args.substring(0, indexOfClosingSquareBracket + 1);
+            args = args.substring(indexOfClosingSquareBracket + 2);
+        }
+
+        Matcher matcher = pattern.matcher(args);
+        if (!matcher.matches()) {
+            System.err.println("Usage: -javaagent:/path/to/JavaAgent.jar=[host:]<port>:<yaml configuration file>");
+            System.exit(1);
+        }
+
+        String givenHost = matcher.group(2);
+        String givenPort = matcher.group(3);
+        String givenConfigFile = matcher.group(4);
+
+        int port = Integer.parseInt(givenPort);
+
+        InetSocketAddress socket;
+        if (givenHost != null && !givenHost.isEmpty()) {
+            socket = new InetSocketAddress(givenHost, port);
+        }
+        else {
+            socket = new InetSocketAddress(ifc, port);
+            givenHost = ifc;
+        }
+
+        return new Config(givenHost, port, givenConfigFile, socket);
+    }
+
+    static class Config {
+        String host;
+        int port;
+        String file;
+        InetSocketAddress socket;
+
+        Config(String host, int port, String file, InetSocketAddress socket) {
+            this.host = host;
+            this.port = port;
+            this.file = file;
+            this.socket = socket;
+        }
+    }
+
+    private static Pattern pattern = Pattern.compile(
+            "^(([\\w.]+):)?" +  // optional host portion (not ipv6)
+            "(\\d{1,5})" +      // port
```

Comment:
Ports can be letters.

I'd suggest counting the colons, and if there's > 2, parsing them from left to right and what's left is the filename.

Reply:
Hmm? Can you clarify when a port can be letters? I'm not familiar with this concept.

Reply:
`http` is the same as 80, per /etc/services.

Reply:
Ah yes, I see. However I'm not sure that's possible in this code base since we are in Java. Java does not have an API to find ports by name. All the APIs that are used in this code base require a port to be an integer. For example all constructor calls to `InetSocketAddress`. Maybe that part can be added as a new feature down the road?

Reply:
So it is.

This code still isn't quite right. Someone can do [ipv6]:host:1234:path

Reply:
If you are talking about ipv6 addresses, I believe that functionality is captured. I left a Github comment on the unit test where this is demonstrated. That functionality was introduced by a previous pull request, and I left that logic as-is. That logic remains at the top of the `parseConfig` method. I tried to make these changes non-invasive.

Reply:
I'm talking specifically about the invalid input `[ipv6]:host:1234:path`, which should be rejected.

> I tried to make these changes non-invasive.

It's more important that it works, and the end code is clean.

Reply:
Having `[ipv6]:host:1234:path` as accepted input was an existing issue in the codebase, but I did go ahead and fix it 😁. I also cleaned up the solution to use a single regex, instead of having separate logic for the ipv6 address.